src/ModelManager.ts

```diff
@@ -125,96 +153,170 @@ export class ModelManager {
-                            });
-                        } else {
-                            const data = this.modelStore.getData();
+        triggerPageModelLoaded(data);
 
-                            triggerPageModelLoaded(data);
+        return data;
+    }
 
-                            return data;
+    /**
+     * Sets initialization promise to fetch model if root path is available
+     * Also, to be returned on synchronous initialization
+     */
+    private _setInitializationPromise(rootModelPath: string) {
+        const {
+            rootModelURL,
+            currentPathname,
+            sanitizedCurrentPathname,
+            metaPropertyModelUrl
+        } = this._modelPaths;
+
+        this._initPromise = (
+            this._checkDependencies().then(() => {
+                const data = this.modelStore.getData(rootModelPath);
+
+                if (data && (Object.keys(data).length > 0)) {
+                    triggerPageModelLoaded(data);
+
+                    return data;
+                } else if (rootModelURL) {
+                    return this._fetchData(rootModelURL).then((rootModel: Model) => {
+                        try {
+                            this.modelStore.initialize(rootModelPath, rootModel);
+
+                            // Append the child page if the page model doesn't correspond to the URL of the root model
+                            // and if the model root path doesn't already contain the child model (asynchronous page load)
+                            if (!!currentPathname && !!sanitizedCurrentPathname &&
+                                !isPageURLRoot(currentPathname, metaPropertyModelUrl) &&
+                                !hasChildOfPath(rootModel, currentPathname)) {
+                                return this._fetchData(currentPathname).then((model: Model) => {
+                                    this.modelStore.insertData(sanitizedCurrentPathname, model);
+
+                                    return this._fetchPageModelFromStore();
+                                });
+                            } else if (!PathUtils.isBrowser()) {
+                                throw new Error(`Attempting to retrieve model data from a non-browser.
+                                    Please provide the initial data with the property key model`
+                                );
+                            }
+
+                            return this._fetchPageModelFromStore();
+                        } catch (e) {
+                            console.error(`Error on initialization - ${e}`);
                         }
-                    } else {
-                        throw new Error('Attempting to retrieve model data from a non-browser. Please provide the initial data with the property key model');
-                    }
-                });
-            }
-        });
-
-        return this._initPromise;
+                    });
+                }
+            })
+        );
```

Comment:
that's quite long, i think we should break down into a smaller functions

Reply:
wdyt @sharanyavinod? :)

Reply:
I've already updated it to reduce the complexity. There is no more tslint warnings for complexity on this, so I think it's good to go. Please take a look @grubyak :)